### PR TITLE
CRYP-203: edit account information, name, email, password

### PR DESCRIPTION
**Pull request by Andre-Ibrahim**

### CODE REVIEW

File: packages/api/src/services/authentication.service.ts [hunk 1]
```diff
@@ -33,6 +37,16 @@ export class AuthenticationService {
         return this.signToken(user);
     }
 
+    async isMyPassword(password: string, userId: number): Promise<string> {
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts]:
```suggestion
    async verify(password: string, userId: number): Promise<string> {
```

File: packages/api/src/services/authentication.service.ts [hunk 2]
```diff
@@ -33,6 +37,16 @@ export class AuthenticationService {
...
+        const user = await this.usersService.findOneById(userId);
+
+        return await bcrypt.compare(password, user.password);
+    }
+
+    async createNewPassword(password: string): Promise<string> {
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:46]:
```suggestion
    async encode(password: string): Promise<string> {
```

File: packages/client/components/list/MultiLineListItem.tsx [hunk 3]
```diff
@@ -29,7 +29,7 @@ export default function MultiLineListItem({
             </Text>
             {altValue && <Text flex={1}>{altValue}</Text>}
             <HStack space="10px">
-                <Text flex={1}>{value}</Text>
```
Comment (jason-gerard) [inline packages/client/components/list/MultiLineListItem.tsx:32]: Why did you change this? Effects a lot of other places

File: packages/client/screens/account/AccountEmailScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                formikHelpers.setFieldError("confirmEmail", error.message);
+            }
+        }
+    }
+
+    return (
+        <View style={styles.view}>
+            <Formik initialValues={intitialValues} validationSchema={updateUserSchema} onSubmit={handleUpdate}>
+                {({ values, errors, touched, handleChange, submitForm }) => (
+                    <VStack space={4} marginTop={5}>
```
Comment (jason-gerard) [inline packages/client/screens/account/AccountEmailScreen.tsx:56]: These dont reference pixel values btw you need to do `space={"4px"}` if that was what you wanted

File: packages/client/screens/account/AccountNameScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                                disabled={
+                                    intitialValues.firstName === values.firstName &&
+                                    intitialValues.lastName === values.lastName
+                                }
+                                style={
+                                    intitialValues.firstName === values.firstName &&
+                                    intitialValues.lastName === values.lastName
+                                        ? styles.ButtonDisabled
+                                        : styles.Button
+                                }
```
Comment (jason-gerard) [inline packages/client/screens/account/AccountNameScreen.tsx:81]: You aren't using the right props, this is how it should be. The disabled styles are already defined

```suggestion
                                isDisabled={
                                    intitialValues.firstName === values.firstName &&
                                    intitialValues.lastName === values.lastName
                                }
                                style={styles.Button}
```

File: packages/common/src/validations/update_user_schema.ts [hunk 6]
```diff
@@ -5,18 +5,19 @@ export const updateUserSchema = yup.object({
...
         .string()
-        .matches(new RegExp("^[A-Za-z]+"), "First name must only contain alphabetic characters.")
+        .matches(new RegExp("^[A-Za-z]+$"), "First name must only contain alphabetic characters.")
         .min(0)
         .max(32, "First name must be 32 characters or less.")
         .nullable(),
     lastName: yup
         .string()
-        .matches(new RegExp("^[A-Za-z]+"), "Last name must only contain alphabetic characters.")
+        .matches(new RegExp("^[A-Za-z]+$"), "Last name must only contain alphabetic characters.")
```
Comment (jason-gerard) [inline packages/common/src/validations/update_user_schema.ts:14]: What is this?

File: packages/api/src/services/users.service.ts [hunk 7]
```diff
@@ -53,12 +56,18 @@ export class UsersService {
...
+                throw new ForbiddenException(ERROR_CURRENT_PASSWORD_INCORRECT);
+            }
+            user.password = await this.authService.createNewPassword(updateUserRequest.newPassword);
         }
         if (updateUserRequest.areNotificationsEnabled != null) {
             user.areNotificationsEnabled = updateUserRequest.areNotificationsEnabled;
         }
+        if (updateUserRequest.email != null) {
+            user.email = updateUserRequest.email;
+        }
```
Comment (jason-gerard) [inline packages/api/src/services/users.service.ts:70]: Confirm email?

File: packages/common/src/validations/update_user_schema.ts (hunk 6 above)
Comment (Andre-Ibrahim) [reply packages/common/src/validations/update_user_schema.ts:14]: it will still match the first part of the string like if you had Jason! it would match Jason, this makes sure that it validates the entire word

File: packages/api/src/services/users.service.ts (hunk 7 above)
Comment (Andre-Ibrahim) [reply packages/api/src/services/users.service.ts:72]: Confirm email is handle by the schema validation same as confirm password

File: packages/common/src/validations/update_user_schema.ts (hunk 6 above)
Comment (jason-gerard) [reply packages/common/src/validations/update_user_schema.ts:14]: Why do we want that?

File: packages/client/components/list/MultiLineListItem.tsx (hunk 3 above)
Comment (Andre-Ibrahim) [reply packages/client/components/list/MultiLineListItem.tsx:32]: the text gets wrapped in weird way over 3 characters with it

File: packages/api/src/services/users.service.ts (hunk 7 above)
Comment (jason-gerard) [reply packages/api/src/services/users.service.ts:72]: That's pretty cool boy

File: packages/client/components/list/MultiLineListItem.tsx (hunk 3 above)
Comment (jason-gerard) [reply packages/client/components/list/MultiLineListItem.tsx:32]: But is it going to mess up the styling in all the other places it is used?

File: packages/common/src/validations/update_user_schema.ts (hunk 6 above)
Comment (Andre-Ibrahim) [reply packages/common/src/validations/update_user_schema.ts:14]: without it it would still pass names with special characters because it matches something

File: packages/client/components/list/MultiLineListItem.tsx (hunk 3 above)
Comment (Andre-Ibrahim) [reply packages/client/components/list/MultiLineListItem.tsx:32]: will take a look before merging the other pages

**Review by PolaFarid — APPROVED** [review]

File: packages/client/components/list/MultiLineListItem.tsx (hunk 3 above)
Comment (Andre-Ibrahim) [reply packages/client/components/list/MultiLineListItem.tsx:32]: you where right i put a minWidth on it so it doesn't look weird instead and doubled checked all pages that use it which looked fine
